packages/server/src/core/bank_rule/specialization_groups.rs

```diff
@@ -47,19 +47,41 @@ fn is_valid_assignment(
     true
 }
 
+fn get_best_match(
+    course_id_to_sg_index: &HashMap<CourseId, usize>,
+    number_of_required_groups: usize,
+) -> HashMap<CourseId, usize> {
+    let mut maybe_best_match = course_id_to_sg_index.to_owned();
+    maybe_best_match.retain(|_, sg_index| {
+        // Retain only the sgs indices which appear at least X times (most of the times, X == 3)
+        course_id_to_sg_index
+            .values()
+            .filter(|group| *group == sg_index)
+            .count()
+            >= number_of_required_groups
+    });
+    maybe_best_match
+}
+
 // This function is looking for a valid assignment for the courses which fulfill the sgs requirements
 // If an assignment is found it returns it, None otherwise.
 fn find_valid_assignment_for_courses(
     sgs: &[SpecializationGroup],
     groups_indices: &[usize],
     optional_sgs_for_course: &HashMap<CourseId, Vec<usize>>, // list of all optional sgs for each course
+    current_best_match: &mut HashMap<CourseId, usize>,       // the best match of sgs
     course_id_to_sg_index: &mut HashMap<CourseId, usize>,
     course_index: usize, // course_index-th element in optional_sgs_for_course
 ) -> Option<HashMap<CourseId, usize>> {
     if course_index >= optional_sgs_for_course.len() {
         if is_valid_assignment(sgs, groups_indices, course_id_to_sg_index) {
             return Some(course_id_to_sg_index.clone());
         }
+        let maybe_best_match = get_best_match(course_id_to_sg_index, groups_indices.len());
+        if maybe_best_match.len() > current_best_match.len() {
```

Comment:
You compare the number of courses and not the number of specialization groups.
Theoretically it may be that the user has more courses but in the same group and therefore it shouldn't be considered as a better match.
A better match should be if the user completed more sgs.